### PR TITLE
WIP - Fix #10865 - long/slow bin parsing issues

### LINKED ISSUE #10865: Slow to analyze some crafted ELF, Mach-o and PE
### Work environment

| Questions                                            | Answers
|------------------------------------------------------|--------------------
| OS/arch/bits (mandatory)                             | Ubuntu x86 64
| File format of the file you reverse (mandatory)      | ELF, Mach-o, MS OMF
| Architecture/bits of the file (mandatory)            | -
| r2 -v full output, **not truncated** (mandatory)     | radare2 2.8.0-git 18863 @ linux-x86-64 git.2.7.0-192-g72fd97738 commit: 72fd9773844d416197ec0fa99df5fbf338a2e749 build: 2018-07-31__11:35:08

### Expected behavior
r2 exits.

### Actual behavior
r2 hangs (different from #10444, which can quit by pressing `q`, these inputs make r2 hangs there, which can be verified by running `r2 $FILE`).

### Steps to reproduce the behavior 
- Download one of the files inside [this directory](https://github.com/ntu-sec/pocs/tree/master/radare2-18863/hangs);
- run `r2 -Q $FILE`.

### PR DESCRIPTION
**Pull request by radare**

### CODE REVIEW

**Comment by radare** [comment]
* [x] hang_01
* [x] hang_02
* [ ] hang_03
* [x] hang_04
* [ ] hang_05
* [ ] hang_06
* [ ] hang_07
* [ ] hang_08
* [ ] hang_09
* [ ] hang_10

**Comment by ret2libc** [comment]
Do you have an ELF as example for this?

**Comment by ret2libc** [comment]
Shall we add those binaries to r2r and reference them in a comment near these fixes? They could help understanding why they were added, since it's immediately clear

File: libr/bin/format/elf/elf.c [hunk 1]
```diff
@@ -3000,6 +3000,17 @@ static RBinElfSymbol* Elf_(_r_bin_elf_get_symbols_imports)(ELFOBJ *bin, int type
 			if (nsym < 0) {
 				goto beach;
 			}
+			{
+				ut64 sh_begin = bin->shdr[i].sh_offset;
+				ut64 sh_end = sh_begin + bin->shdr[i].sh_size;
+				if (sh_begin > bin->size) {
+					goto beach;
+				}
+				if (sh_begin > bin->size || sh_end > bin->size) {
```
[the remark this answers is not in the document]
Comment (radare) [reply libr/bin/format/elf/elf.c]: fixed

**Comment by radare** [comment]
all the fuzzed bins i fix are pushed to r2r right after fixing them. pushing them now makes no sense because it will break travis. I always do a PR or push them directly. see bins/fuzzed. been doing this obvious thing for years already

**Comment by radare** [comment]
the reason for defining better commit messages is to help having proper referencs to which bin, test, issue each commit fixes. use tags and keywords to make them parseable, etc. but this hasnt been defined yet

**Comment by ret2libc** [comment]
okok. seems reasonable.
btw, can you reproduce the hang if you don't use `-Q`? On OS X, I can't :(

**Comment by ret2libc** [comment]
Not hang_01 at least.

**Comment by radare** [comment]
yes i can reproduce all those hangs without -Q, just build with asan on osx. and those are not hangs. its just slow because they are looping way beyond the end of the file, so parsing ffffff and allocating huge pages mainly.

**Comment by radare** [comment]
most of those boundary checks should be done in a different way i think, im worried about fixing them now to make r2 more stable, but im pretty sure there may be better ways to shrink boundaries. gtg take plane now. will be committing things offline now

**Comment by radare** [comment]
We should define that rule. For me it seems reasonable, but we can do that in a separate pr including other static functions in a single commit.

**Comment by radare** [comment]
I mean, updating the coding style guidelines document to specify the naming conventions on static functions because public ones are modulename_method and here would be $verb$type$state “is section invalid” and how to construct other private functions

**Pull request merged**